data-prepper-core/src/main/java/com/amazon/dataprepper/pipeline/ProcessWorker.java

```diff
@@ -54,6 +55,7 @@ public void run() {
                 }
                 if (!records.isEmpty()) {
                     postToSink(records); //TODO use the response to ack the buffer on failure?
+                    readBuffer.checkpoint(new CheckpointState(records.size()));
                 }
```

Comment:
What if this records list is empty, say in the event of Peer Forwarder forwarding all of the records?

Reply:
Great point! I think in this case we do need to apply your idea on passing the Checkpoint obj to consumer.

Reply:
Oh actually, this is a blocker. There is no notion that the number of output records is equal to the number of input records. We should wait for this to complete and checkpoint with the number of records read from the buffer. We don't care how many are output at the end.